### PR TITLE
Add callSite - classLoaderInfo integration

### PR DESCRIPTION
**Pull request by francischempa**
Integration of the classLoaderInfo into CallSite which is now able to lookup sourceFile names. RecordingSetting, OSInformation, InitialSystemProperty removed from the ID mappings due to inconsistencies. The implementation can now handle class names of both types - “com.package.ClassName” as well as “ClassName”.
**NB:** All $$Lambda frames are aggregated as TopGunIgnoredFrame

### CODE REVIEW

File: cmdLine/src/main/scala/topgun/cmdline/ClassLoaderInfo.scala [hunk 1]
```diff
@@ -14,18 +15,22 @@ class ClassLoaderInfo() {
...
+      val parsingFlagNone = 0;
+      reader.accept(classNode, parsingFlagNone)
+      iStream.close()
+      val classNameFromReader = reader.getClassName
+      val sourceFile = classNode.sourceFile
+      val superClass = reader.getSuperName
+      val interfaces: List[String] = reader.getInterfaces.toList
+      ClassInfo(classNameFromReader, superClass, sourceFile, interfaces)
+    }catch {
+      case e: IOException => ClassInfo("className not found","super not found","sourceFile not found",List())
```
Comment (mkeskells) [inline cmdLine/src/main/scala/topgun/cmdline/ClassLoaderInfo.scala]: too general a catch

File: cmdLine/src/main/scala/topgun/cmdline/FileParser.scala [hunk 2]
```diff
@@ -228,9 +227,17 @@ class FileParser(file: File, cmdLine: JfrParseCommandLine, totals: Totals, confi
           val method = frame.getMethod
           if(method ne null) {
             val splitIndex = method.getType.getName.lastIndexOf(".")
+            var packageName:String = ""
+            var methodName:String = ""
+            if(splitIndex != -1) {
```
Comment (mkeskells) [inline cmdLine/src/main/scala/topgun/cmdline/FileParser.scala]: dont use vars

File: cmdLine/src/main/scala/topgun/cmdline/FileParser.scala [hunk 3]
```diff
@@ -228,9 +227,17 @@ class FileParser(file: File, cmdLine: JfrParseCommandLine, totals: Totals, confi
...
+              packageName = method.getType.getName.substring(0, splitIndex).intern()
+              methodName = method.getType.getName.substring(splitIndex + 1).intern()
+            }else {
+              methodName = method.getType.getName
+            }
             CallSite(
-              method.getType.getName.substring(0,splitIndex).intern(),
-              method.getType.getName.substring(splitIndex+1).intern(),
+              packageName.intern(),
```
Comment (mkeskells) [inline cmdLine/src/main/scala/topgun/cmdline/FileParser.scala]: why are you interning

File: cmdLine/src/main/scala/topgun/cmdline/ClassLoaderInfo.scala [hunk 4]
```diff
@@ -14,18 +15,22 @@ class ClassLoaderInfo() {
...
-    val interfaces:List[String] = reader.getInterfaces.toList
-    ClassInfo(classNameFromReader, superClass, sourceFile, interfaces)
+    try {
+      val resourceName = className.replace(".", "/") + ".class"
+      val iStream = classLoader.getResourceAsStream(resourceName)
+      val reader = new ClassReader(iStream)
+      val classNode = new ClassNode()
+      // specify no parsing options.
+      val parsingFlagNone = 0;
+      reader.accept(classNode, parsingFlagNone)
```
Comment (mkeskells) [inline cmdLine/src/main/scala/topgun/cmdline/ClassLoaderInfo.scala]: may as well pass 0 to the method

File: cmdLine/src/main/scala/topgun/cmdline/ClassLoaderInfo.scala (hunk 1 above)
Comment (francischempa) [reply cmdLine/src/main/scala/topgun/cmdline/ClassLoaderInfo.scala]: The ASM ClassReader constructor throws that IOException and the message in there is class not found. 
public ClassReader(final String className) throws IOException

File: cmdLine/src/main/scala/topgun/cmdline/ClassLoaderInfo.scala (hunk 4 above)
Comment (francischempa) [reply cmdLine/src/main/scala/topgun/cmdline/ClassLoaderInfo.scala]: done

File: cmdLine/src/main/scala/topgun/cmdline/FileParser.scala (hunk 3 above)
Comment (francischempa) [reply cmdLine/src/main/scala/topgun/cmdline/FileParser.scala]: We don't want different copies of that particular string to exist. Allow of callSite could have the same packageName, className but different line numbers.we keep one of that string in memory

File: cmdLine/src/main/scala/topgun/cmdline/ClassLoaderInfo.scala (hunk 1 above)
Comment (mkeskells) [reply cmdLine/src/main/scala/topgun/cmdline/ClassLoaderInfo.scala]: and in that case, what is the InputStream?

File: cmdLine/src/main/scala/topgun/cmdline/FileParser.scala (hunk 3 above)
Comment (mkeskells) [reply cmdLine/src/main/scala/topgun/cmdline/FileParser.scala]: look at your code, and work out what you need to intern, and where

File: cmdLine/src/main/scala/topgun/cmdline/FileParser.scala [hunk 5]
```diff
@@ -228,9 +227,17 @@ class FileParser(file: File, cmdLine: JfrParseCommandLine, totals: Totals, confi
           val method = frame.getMethod
           if(method ne null) {
             val splitIndex = method.getType.getName.lastIndexOf(".")
+            var packageName:String = ""
+            var methodName:String = ""
```
Comment (mkeskells) [inline cmdLine/src/main/scala/topgun/cmdline/FileParser.scala]: it this a method or a class name?

File: cmdLine/src/main/scala/topgun/core/CallSite.scala [hunk 6]
```diff
@@ -37,10 +40,13 @@ class CallSite private(val packageName:String, val className: String, val method
   def isIgnorableTopFrame = (flags & CallSite.IGNORABLE_TOP_FRAME) != 0
 
   lazy val filename = {
-    val i = className.indexOf('$')
-    if (i == -1) className else className.substring(0, i)
+    def getResourceName(className:String, packageName:String): String = {
```
Comment (mkeskells) [inline cmdLine/src/main/scala/topgun/core/CallSite.scala]: why is this a method, not a val

File: cmdLine/src/main/scala/topgun/cmdline/FileParser.scala (hunk 2 above)
Comment (francischempa) [reply cmdLine/src/main/scala/topgun/cmdline/FileParser.scala]: done

File: cmdLine/src/main/scala/topgun/cmdline/FileParser.scala (hunk 5 above)
Comment (francischempa) [reply cmdLine/src/main/scala/topgun/cmdline/FileParser.scala]: it is a className. fixed

File: cmdLine/src/main/scala/topgun/core/CallSite.scala (hunk 6 above)
Comment (francischempa) [reply cmdLine/src/main/scala/topgun/core/CallSite.scala]: fixed now

File: cmdLine/src/main/scala/topgun/cmdline/ClassLoaderInfo.scala (hunk 1 above)
Comment (francischempa) [reply cmdLine/src/main/scala/topgun/cmdline/ClassLoaderInfo.scala]: fixed. 
InputStream was null

File: cmdLine/src/main/scala/topgun/cmdline/FileParser.scala (hunk 3 above)
Comment (francischempa) [reply cmdLine/src/main/scala/topgun/cmdline/FileParser.scala]: found it.

**Comment by mkeskells** [comment]
@retronym commented The synthetic lambda apply method will call the lambda implementation method with the actual code; that method has line number info.

so we should be able to just ignore these "$$Lambda" class frames

File: cmdLine/src/main/scala/topgun/cmdline/FileParser.scala [hunk 7]
```diff
@@ -228,9 +227,11 @@ class FileParser(file: File, cmdLine: JfrParseCommandLine, totals: Totals, confi
           val method = frame.getMethod
           if(method ne null) {
             val splitIndex = method.getType.getName.lastIndexOf(".")
+            val packageName = method.getType.getName.substring(0, splitIndex).intern()
```
Comment (mkeskells) [inline cmdLine/src/main/scala/topgun/cmdline/FileParser.scala]: logic on `splitIndex` should be here

e.g val packageName  if (splitIndex == -1)....

**Review by mkeskells — CHANGES_REQUESTED** [review]

File: cmdLine/src/main/scala/topgun/cmdline/ClassLoaderInfo.scala (hunk 1 above)
Comment (mkeskells) [reply cmdLine/src/main/scala/topgun/cmdline/ClassLoaderInfo.scala]: You marked this as resolved, but the code is unchanged

**Comment by francischempa** [comment]
> @retronym commented The synthetic lambda apply method will call the lambda implementation method with the actual code; that method has line number info.
> 
> so we should be able to just ignore these "$$Lambda" class frames

will these frame be ignored from the command line  or should we add it pragmatically.

File: cmdLine/src/main/scala/topgun/cmdline/ClassLoaderInfo.scala (hunk 1 above)
Comment (francischempa) [reply cmdLine/src/main/scala/topgun/cmdline/ClassLoaderInfo.scala]: It has changed. I placed a check to see if the input stream is null before continuing to the ClassReader.

File: cmdLine/src/main/scala/topgun/cmdline/FileParser.scala [hunk 8]
```diff
@@ -228,18 +227,25 @@ class FileParser(file: File, cmdLine: JfrParseCommandLine, totals: Totals, confi
...
-              method.getName.intern(),
-              method.getDescriptor.intern(),
-              frame.getLineNumber,
-              classPath.intern()
-            )
+            val packageName = if (splitIndex != -1) method.getType.getName.substring(0, splitIndex).intern() else ""
+            val className = if (splitIndex != -1) method.getType.getName.substring(splitIndex + 1).intern() else method.getType.getName.intern()
+
+            if(packageName.contains("$$Lambda")){
+              CallSite("TopGunIgnoredFrame", "TopGunIgnoredFrame", "TopGunIgnoredFrame", "TopGunIgnoredFrame", 0,classPath)
```
Comment (mkeskells) [inline cmdLine/src/main/scala/topgun/cmdline/FileParser.scala]: these should be filtered not aggregated, and add comment to indicate where they came from

File: cmdLine/src/main/scala/topgun/cmdline/FileParser.scala [hunk 9]
```diff
@@ -228,18 +227,30 @@ class FileParser(file: File, cmdLine: JfrParseCommandLine, totals: Totals, confi
...
-              frame.getLineNumber,
-              classPath.intern()
-            )
+            val packageName = if (splitIndex != -1) method.getType.getName.substring(0, splitIndex).intern() else ""
+            val className = if (splitIndex != -1) method.getType.getName.substring(splitIndex + 1).intern() else method.getType.getName.intern()
+
+            // The synthetic lambda apply method will call the lambda implementation method with the actual code; that method has line number info.
+            // so we should be able to just ignore these "$$Lambda" class frames.
+            // map all lambda frames to null and filter out after map
+            if(packageName.contains("$$Lambda")){
```
Comment (retronym) [inline cmdLine/src/main/scala/topgun/cmdline/FileParser.scala]: Alternatively, `frame.getMethod.isHidden`.

File: cmdLine/src/main/scala/topgun/cmdline/RecordedEventLabelMapping.scala [hunk 10]
```diff
@@ -0,0 +1,13 @@
...
+// TODO find reference values inside JMC/JFR and replace with constants.
+//  RecrodedEvent.AllocatedInNewTLAB
+object RecordedEventLabelMapping {
+  val AllocationInNewTLAB = "Allocation in new TLAB"
+  val AllocationOutsideTlab = "Allocation outside TLAB"
+  val MethodProfilingSample = "Method Profiling Sample"
+  val MethodProfilingSampleNative = "Method Profiling Sample Native"
+  val InitialSystemProperty = "Initial System Property"
+  val RecordingSetting = "Recording Setting"
+  val OSInformation = "OS Information"
```
Comment (retronym) [inline cmdLine/src/main/scala/topgun/cmdline/RecordedEventLabelMapping.scala]: I think that its better to switch on `(event: Event) => e.getEventType().getType().getName()`, which has ids like `jdk.ObjectAllocationInNewTLAB`. (full list: https://github.com/openjdk/jdk/blob/master/src/hotspot/share/jfr/metadata/metadata.xml)

File: cmdLine/src/main/scala/topgun/cmdline/FileParser.scala [hunk 11]
```diff
@@ -228,18 +227,30 @@ class FileParser(file: File, cmdLine: JfrParseCommandLine, totals: Totals, confi
...
-              frame.getLineNumber,
-              classPath.intern()
-            )
+            val packageName = if (splitIndex != -1) method.getType.getName.substring(0, splitIndex).intern() else ""
+            val className = if (splitIndex != -1) method.getType.getName.substring(splitIndex + 1).intern() else method.getType.getName.intern()
+
+            // The synthetic lambda apply method will call the lambda implementation method with the actual code; that method has line number info.
+            // so we should be able to just ignore these "$$Lambda" class frames.
+            // map all lambda frames to null and filter out after map
+            if(frame.getMethod.isHidden){
```
Comment (mkeskells) [inline cmdLine/src/main/scala/topgun/cmdline/FileParser.scala]: formatting

File: cmdLine/src/main/scala/topgun/cmdline/FileParser.scala [hunk 12]
```diff
@@ -228,18 +227,30 @@ class FileParser(file: File, cmdLine: JfrParseCommandLine, totals: Totals, confi
...
+                frame.getLineNumber,
+                classPath.intern()
+              )
+            }
           } else {
             CallSite("NoMethodInFrame_", "NoMethodInFrame_", "NoMethodInFrame_", "NoMethodInFrame_", frame.getLineNumber,classPath.intern())
           }
-      }.distinct.takeWhile { f => !f.isIgnorableTopFrame }.toList
+      // $Lambda frames are mapped to null and filtered out.
+      }.filter(f => f ne null)
```
Comment (mkeskells) [inline cmdLine/src/main/scala/topgun/cmdline/FileParser.scala]: use _ for simple expressions
`
```suggestion
      }.filter(_ ne null)
```

File: cmdLine/src/main/scala/topgun/cmdline/ClassLoaderInfo.scala [hunk 13]
```diff
@@ -16,16 +18,18 @@ class ClassLoaderInfo() {
...
-    // specify no parsing options.
-    val parsingFlagNone = 0;
-    reader.accept(classNode, parsingFlagNone)
-    iStream.close()
-    val classNameFromReader = reader.getClassName
-    val sourceFile = classNode.sourceFile
-    val superClass = reader.getSuperName
-    val interfaces:List[String] = reader.getInterfaces.toList
-    ClassInfo(classNameFromReader, superClass, sourceFile, interfaces)
+    if(iStream ne null) {
```
Comment (mkeskells) [inline cmdLine/src/main/scala/topgun/cmdline/ClassLoaderInfo.scala]: formatting

**Pull request merged**